src/Raven.Server/Documents/TransactionOperationsMerger.cs

```diff
@@ -827,34 +874,23 @@ private enum PendingOperations
 
                 var llt = context.Transaction.InnerTransaction.LowLevelTransaction;
 
-                if (_parent.Configuration.Memory.EnableHighTemporaryDirtyMemoryUse)
+                var dirtyMemoryState = LowMemoryNotification.Instance.DirtyMemoryState;
+                if (dirtyMemoryState.IsHighDirty)
                 {
                     var now = _parent.Time.GetUtcNow();
-                    if (now - _lastHighDirtyMemCheck > _timeToCheckHighDirtyMemory.AsTimeSpan) // we do not need to test scratch dirty mem every write
+                    if (now - _lastHighDirtyMemCheck > _timeToCheckHighDirtyMemory.AsTimeSpan)
                     {
-                        if (MemoryInformation.IsHighDirtyMemory(percentageFromPhysicalMem, out var details))
-                        {
-                            var highDirtyMemory = new HighDirtyMemoryException(
-                                $"Operation was cancelled by the transaction merger for transaction #{llt.Id} due to high dirty memory in scratch files." +
-                                $" This might be caused by a slow IO storage. Current memory usage: {details}");
-
-                            foreach (var pendingOp in pendingOps)
-                                pendingOp.Exception = highDirtyMemory;
-
-                            NotifyOnThreadPool(pendingOps);
-
-                            var rejectedBuffer = GetBufferForPendingOps();
-                            while (_operations.TryDequeue(out var operationToReject))
-                            {
-                                operationToReject.Exception = highDirtyMemory;
-                                rejectedBuffer.Add(operationToReject);
-                            }
-                            NotifyOnThreadPool(rejectedBuffer);
-                            GlobalFlushingBehavior.GlobalFlusher.Value?.MaybeFlushEnvironment(context.Environment); // commit stage 2 won't trigger so we need to ask for flush here
-                            break;
-                        }
-                        _lastHighDirtyMemCheck = now; // reset timer for next check only if no errors (otherwise check every single write until back to normal)
+                        // we need to ask for a flush here
+                        GlobalFlushingBehavior.GlobalFlusher.Value?.MaybeFlushEnvironment(context.Environment);
+                        _lastHighDirtyMemCheck = now;
                     }
+
```

Comment:
Suggestion: Maybe not throw immediately here, because we request a flush and it might be fast enough. So we can wait longer and just then throw.

Reply:
Flush isn't actually going to release any memory, the OS still need to do some work around that.
Would rather trigger the flush and wait for the _next_ operation

Reply:
By "wait longer" I meant: raise flag and if IsHighDirty is still on after X seconds, start rejecting.
And it makes more sense to start rejecting after period of time and not after "operation" as it is disk write speeds issue.